snuba/clickhouse/astquery.py

```diff
@@ -0,0 +1,163 @@
+        self.__having = query.get_having_from_ast()
+        self.__orderby = query.get_orderby_from_ast()
+        self.__data_source = query.get_data_source()
+        self.__arrayjoin = query.get_arrayjoin_from_ast()
+        self.__granularity = query.get_granularity()
+        self.__limit = query.get_limit()
+        self.__limitby = query.get_limitby()
+        self.__offset = query.get_offset()
+
+        # Clickhouse specific fields
+        self.__prewhere = query.get_prewhere_ast()
+        self.__turbo = settings.get_turbo()
+        self.__final = query.get_final()
+
+        # Attributes that should be clickhouse specific that
+        # have to be removed from the Snuba Query
+        self.__sample = query.get_sample()
+        self.__hastotals = query.has_totals()
+
+        self.__settings = settings
+        self.__formatted_query: Optional[str] = None
+
+    def format_sql(self) -> str:
+        if self.__formatted_query:
+            return self.__formatted_query
+
+        parsing_context = ParsingContext()
+
+        selected_cols = [
+            e.accept(ClickhouseExpressionFormatter(parsing_context))
+            for e in self.__selected_columns
+        ]
+        formatted_columns = ", ".join(selected_cols)
+        select_clause = f"SELECT {formatted_columns}"
+
+        # TODO: The visitor approach will be used for the FROM clause as well.
+        from_clause = f"FROM {self.__data_source.format_from()}"
+
+        if self.__final:
+            from_clause = f"{from_clause} FINAL"
+
+        # TODO: Sampling rate will become one step of Clickhouse query processing
+        if not self.__data_source.supports_sample():
+            sample_rate = None
+        else:
+            if self.__sample:
+                sample_rate = self.__sample
+            elif self.__settings.get_turbo():
+                sample_rate = settings.TURBO_SAMPLE_RATE
+            else:
+                sample_rate = None
+        if sample_rate:
+            from_clause = f"{from_clause} SAMPLE {sample_rate}"
+
+        join_clause = ""
+        if self.__arrayjoin:
+            formatted_array_join = self.__arrayjoin.accept(
+                ClickhouseExpressionFormatter(parsing_context)
+            )
+            join_clause = f"ARRAY JOIN {formatted_array_join}"
+
+        where_clause = ""
+        if self.__condition:
+            formatted_condition = self.__condition.accept(
+                ClickhouseExpressionFormatter(parsing_context)
+            )
+            where_clause = f"WHERE {formatted_condition}"
+
+        # TODO: Pre where processing will become a step in Clickhouse Query processing
+        # instead of being pulled from the Snuba Query
+        prewhere_clause = ""
+        if self.__prewhere:
+            formatted_prewhere = self.__prewhere.accept(
+                ClickhouseExpressionFormatter(parsing_context)
+            )
+            prewhere_clause = f"PREWHERE {formatted_prewhere}"
```

Comment:
It probably makes sense to flip the order of prewhere and where processing here to align with the ordering of the clauses themselves, unless there's some reason they have to be processed in this order.

Reply:
ok